A16-tcp-user-timeout.md

```diff
@@ -0,0 +1,55 @@
+TCP User Timeout
+----
+* Author(s): Yash Tibrewal
+* Approver:
+* Status: Draft
+* Implemented in:
+* Last updated: 2018-8-22
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/6VZYCFZpyTI
+
+## Abstract
+
+Provide an option to set the socket option, TCP_USER_TIMEOUT on TCP sockets for Linux kernels 2.6.37 and later.
+
+## Background
+
+TCP_USER_TIMEOUT is a socket option that has been available on Linux since kernel version 2.6.37. As described the Linux manual page, "This option takes an unsigned int as an argument. When the value is greater than 0, it specifies the maximum amount of time in milliseconds that transmitted data may remain unacknowledged before TCP will forcibly close the corresponding connection and return ETIMEDOUT to the application.  If the option value is specified as 0, TCP will to use the system default."
```

Comment:
given that one of the issues referenced is a python issue, has this been verified to work on manylinux?

Reply:
It will depend on the linux kernel version being used.. if it's lower than 2.6.37 then it will not work

Reply:
manylinux is older than 2.6.37. That said, you can still support the newer kernels and older kernels in the same binary. But if you're on the older machines this won't do anything.

Reply:
CentOS 5 is on 2.6.18. CentOS 6 is 2.6.32. Based on the dockerfile at https://github.com/pypa/manylinux, it seems manylinux implies CentOS/RHEL 5. I'll note that CentOS 5 is unsupported, but RHEL 5 is in "extended support".

Reply:
The manylinux case isn't what you seem to talk about. Manylinux is only the building platform, not the running platform. But if the building platform doesn't have the TCP_USER_TIMEOUT symbol defined, the code to toggle it won't be compiled in, and it won't work, even in the case of the binary being run on a newer platform.

Reply:
It might be relevant to link to the current pull request that's implementing this proposal :-)

https://github.com/grpc/grpc/pull/16419/files

Reply:
It's linked to in "implementation".